### PR TITLE
[SPARK-23271[SQL] Parquet output contains only _SUCCESS file after writing an empty dataframe

### PR DESCRIPTION
**Pull request by dilipbiswal**
## What changes were proposed in this pull request?
Below are the two cases. 
``` SQL
case 1

scala> List.empty[String].toDF().rdd.partitions.length
res18: Int = 1
```
When we write the above data frame as parquet, we create a parquet file containing
just the schema of the data frame. 

Case 2
``` SQL

scala> val anySchema = StructType(StructField("anyName", StringType, nullable = false) :: Nil)
anySchema: org.apache.spark.sql.types.StructType = StructType(StructField(anyName,StringType,false))
scala> spark.read.schema(anySchema).csv("/tmp/empty_folder").rdd.partitions.length
res22: Int = 0
```
For the 2nd case, since number of partitions = 0, we don't call the write task (the task has logic to create the empty metadata only parquet file)

The fix is to create a dummy single partition RDD and set up the write task based on it to ensure
the metadata-only file.

## How was this patch tested?

A new test is added to DataframeReaderWriterSuite.

### CODE REVIEW

**Comment by gatorsmile** [comment]
Update the title to `[SPARK-23271] [SQL] Parquet output contains only _SUCCESS file after writing an empty dataframe`

**Comment by gatorsmile** [comment]
cc @cloud-fan @zsxwing

File: sql/core/src/test/scala/org/apache/spark/sql/test/DataFrameReaderWriterSuite.scala [hunk 1]
```diff
@@ -301,7 +301,6 @@ class DataFrameReaderWriterSuite extends QueryTest with SharedSQLContext with Be
       intercept[AnalysisException] {
         spark.range(10).write.format("csv").mode("overwrite").partitionBy("id").save(path)
       }
-      spark.emptyDataFrame.write.format("parquet").mode("overwrite").save(path)
     }
```
Comment (dilipbiswal) [inline sql/core/src/test/scala/org/apache/spark/sql/test/DataFrameReaderWriterSuite.scala]: Its not legal to write an empty struct in parquet.  Its explained by Herman in [SPARK-20593](https://issues.apache.org/jira/browse/SPARK-20593). Previously, we didn't setup a write 
task for this where as now with this fix we do.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/FileFormatWriterSuite.scala [hunk 2]
```diff
@@ -32,6 +33,24 @@ class FileFormatWriterSuite extends QueryTest with SharedSQLContext {
...
+      withTempPath { outputPath =>
+        val anySchema = StructType(StructField("anyName", StringType) :: Nil)
+        val df = spark.read.schema(anySchema).csv(inputPath.toString)
+        df.write.parquet(outputPath.toString)
+        val partFiles = outputPath.listFiles()
+          .filter(f => f.isFile && !f.getName.startsWith(".") && !f.getName.startsWith("_"))
+        assert(partFiles.length === 1)
+
+        // Now read the file.
+        val  df1 = spark.read.parquet(outputPath.toString)
```
Comment (gatorsmile) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/FileFormatWriterSuite.scala]: Nit: extra space before `df1`

Comment (dilipbiswal) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/FileFormatWriterSuite.scala]: @gatorsmile Thank you. Fixed.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/FileFormatWriter.scala [hunk 3]
```diff
@@ -190,9 +190,13 @@ object FileFormatWriter extends Logging {
           global = false,
           child = plan).execute()
       }
-      val ret = new Array[WriteTaskResult](rdd.partitions.length)
+
+      // SPARK-23271 If we are attempting to write a zero partition rdd, change the number of
+      // partition to 1 to make sure we at least set up one write task to write the metadata.
+      val finalRdd = if (rdd.partitions.length == 0) rdd.repartition(1) else rdd
```
Comment (pashazm) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/FileFormatWriter.scala]: Looks like shuffle will be here if partitions number is zero. If so, maybe, other solution is possible?

Comment (cloud-fan) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/FileFormatWriter.scala]: Yea the shuffle can be avoided. We can just launch a write task for empty RDD, instead of calling `rdd.repartition(1)`.

Comment (dilipbiswal) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/FileFormatWriter.scala]: @cloud-fan @pashazm I was thinking, this would not be a regular event to write empty datasets , right ? Should we be even optimizing this path ? Secondly, is shuffling an empty data set that expensive ?

@cloud-fan, actually i had tried to launch a write task for empty RDD, but was hitting a NullPointerException from scheduler ? Looks like things are setup to only work off of partitions of RDD. Could we try to create this empty metadata file from the driver in this case ? If we go that route, then we may have to refactor the write task code. Seems like a lot for this little corner case, what do you think ?

Comment (hvanhovell) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/FileFormatWriter.scala]: You could try `coalesce(1)` that should not shuffle.

Comment (dilipbiswal) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/FileFormatWriter.scala]: @hvanhovell Thanks. I have a question. Can we go from zero partition to one partition with coalesce() ? In the code we seem to be doing a min(prevPartition, requestedPartition) to set the target number of partition [code](
https://github.com/apache/spark/blob/e26dac5feb02033f980b1e69c9b0ff50869b6f9e/core/src/main/scala/org/apache/spark/rdd/CoalescedRDD.scala#L393 )

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/FileFormatWriterSuite.scala [hunk 4]
```diff
@@ -32,6 +33,24 @@ class FileFormatWriterSuite extends QueryTest with SharedSQLContext {
     }
   }
 
+  test("SPARK-23271 empty dataframe when saved in parquet should write a metadata only file") {
```
Comment (dongjoon-hyun) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/FileFormatWriterSuite.scala]: If this is specific to `parquet`, can we have this `ParquetFileFormatSuite` instead?

Comment (dilipbiswal) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/FileFormatWriterSuite.scala]: @dongjoon-hyun Thank you. Let me check if we have similar issue for orc. If not, i will move it to ParquetFileFormatSuite.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/FileFormatWriter.scala (hunk 3 above)
Comment (dilipbiswal) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/FileFormatWriter.scala]: @hvanhovell Just tried. We stay at numPartitions = 0 after coalesce(). So it does not fix the problem.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/FileFormatWriterSuite.scala (hunk 4 above)
Comment (dongjoon-hyun) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/FileFormatWriterSuite.scala]: Thank you @dilipbiswal . 
I checked with ORC, too. Your patch work for ORC two.
In this suite, can you extend the test case for ORC too?

Comment (dilipbiswal) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/FileFormatWriterSuite.scala]: Thank you very much @dongjoon-hyun. You are super quick :-). Yes, i will add the test case for ORC.

Comment (dongjoon-hyun) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/FileFormatWriterSuite.scala]: Ur, `FileBasedDataSourceSuite` may be more suitable. It has a similar test case. You can add your test case in the similar manner.
https://github.com/apache/spark/blob/master/sql/core/src/test/scala/org/apache/spark/sql/FileBasedDataSourceSuite.scala#L59-L73

Comment (dilipbiswal) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/FileFormatWriterSuite.scala]: @dongjoon-hyun Sure. Will take a look. Thanks !!

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/FileFormatWriter.scala (hunk 3 above)
Comment (cloud-fan) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/FileFormatWriter.scala]: One simple way to fix it: create an empty 1-partition RDD and use it here.

Comment (jiangxb1987) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/FileFormatWriter.scala]: Yea, you can have
```
sparkSession.sparkContext.parallelize(Array.empty[InternalRow])
```

Comment (dilipbiswal) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/FileFormatWriter.scala]: @cloud-fan @jiangxb1987 Thanks a **LOT**. This works perfectly.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/FileFormatWriter.scala [hunk 5]
```diff
@@ -190,9 +190,18 @@ object FileFormatWriter extends Logging {
           global = false,
           child = plan).execute()
       }
-      val ret = new Array[WriteTaskResult](rdd.partitions.length)
+
+      // SPARK-23271 If we are attempting to write a zero partition rdd, create a dummy single
+      // partition rdd to make sure we at least set up one write task to write the metadata.
+      val finalRdd = if (rdd.partitions.length == 0) {
+        sparkSession.sparkContext.parallelize(Array.empty[InternalRow])
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/FileFormatWriter.scala]: `sparkSession.sparkContext.parallelize(Array.empty[InternalRow], 1)`

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/FileFormatWriterSuite.scala [hunk 6]
```diff
@@ -19,6 +19,7 @@ package org.apache.spark.sql.execution.datasources
 
 import org.apache.spark.sql.{QueryTest, Row}
 import org.apache.spark.sql.test.SharedSQLContext
+import org.apache.spark.sql.types.{StringType, StructField, StructType}
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/FileFormatWriterSuite.scala]: unnecessary change.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/FileFormatWriter.scala [hunk 7]
```diff
@@ -190,9 +190,18 @@ object FileFormatWriter extends Logging {
           global = false,
           child = plan).execute()
       }
-      val ret = new Array[WriteTaskResult](rdd.partitions.length)
+
+      // SPARK-23271 If we are attempting to write a zero partition rdd, create a dummy single
+      // partition rdd to make sure we at least set up one write task to write the metadata.
+      val finalRdd = if (rdd.partitions.length == 0) {
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/FileFormatWriter.scala]: how about `val rddWithNonEmptyPartitions ...`

Comment (dilipbiswal) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/FileFormatWriter.scala]: Sure.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/FileFormatWriterSuite.scala (hunk 6 above)
Comment (dilipbiswal) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/FileFormatWriterSuite.scala]: will remove

**Comment by dilipbiswal** [comment]
retest this please

**Comment by gatorsmile** [comment]
BTW, this is a behavior change. We need to write it in the migration guide.

**Comment by dilipbiswal** [comment]
retest this please

**Comment by dilipbiswal** [comment]
@gatorsmile Thanks. I will create a doc pr and address it.

**Comment by cloud-fan** [comment]
I think it's better to have the doc change in the same PR, then it's more clear which patch caused the behavior change.

**Comment by dilipbiswal** [comment]
@cloud-fan Actually i had already created the doc pr in the morning using the same JIRA number. Whenchen, if we want to have both the changes in the same commit , will we be able to do it when we merge the patch ? If not, pl let me know , i will close that PR and move over the change to this branch.

**Comment by cloud-fan** [comment]
no we can't merge 2 PRs together. Please pick one of your PRs and put all the changes there, thanks!

**Comment by dilipbiswal** [comment]
@cloud-fan @gatorsmile Done.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/FileFormatWriterSuite.scala (hunk 6 above)
Comment (cloud-fan) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/FileFormatWriterSuite.scala]: please remove it

File: sql/core/src/test/scala/org/apache/spark/sql/test/DataFrameReaderWriterSuite.scala [hunk 8]
```diff
@@ -301,7 +301,6 @@ class DataFrameReaderWriterSuite extends QueryTest with SharedSQLContext with Be
       intercept[AnalysisException] {
         spark.range(10).write.format("csv").mode("overwrite").partitionBy("id").save(path)
       }
-      spark.emptyDataFrame.write.format("parquet").mode("overwrite").save(path)
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/test/DataFrameReaderWriterSuite.scala]: How does it fail? If it's a runtime error we should fail earlier during analysis. This worth a new JIRA.

**Comment by zsxwing** [comment]
@tdas @brkyvz Do we still need the fix for 0-partition DataFrame in Structured Streaming after this change?

File: docs/sql-programming-guide.md [hunk 9]
```diff
@@ -1930,6 +1930,9 @@ working with timestamps in `pandas_udf`s to get the best performance, see
     - Literal values used in SQL operations are converted to DECIMAL with the exact precision and scale needed by them.
     - The configuration `spark.sql.decimalOperations.allowPrecisionLoss` has been introduced. It defaults to `true`, which means the new behavior described here; if set to `false`, Spark uses previous rules, ie. it doesn't adjust the needed scale to represent the values and it returns NULL if an exact representation of the value is not possible.
 
+ - Since Spark 2.3, writing an empty dataframe (a dataframe with 0 partitions) in parquet or orc format, creates a format specific metadata only file. In prior versions the metadata only file was not created. As a result, subsequent attempt to read from this directory fails with AnalysisException while inferring schema of the file. For example : df.write.format("parquet").save("outDir")
```
[the remark this answers is not in the document]
Comment (dilipbiswal) [reply docs/sql-programming-guide.md]: even -> even if ?
self-described -> self-describing ?
@cloud-fan Nicely written. Thanks. Let me know if you are ok with the above two change ?

[the remark this answers is not in the document]
Comment (cloud-fan) [reply docs/sql-programming-guide.md]: yea the above 2 changes are good!

[the remark this answers is not in the document]
Comment (dilipbiswal) [reply docs/sql-programming-guide.md]: "launches at least one write task"
Actually isn't it exactly one write task ? I am okay with what you have. Just wanted to check to make sure.

File: sql/core/src/test/scala/org/apache/spark/sql/test/DataFrameReaderWriterSuite.scala (hunk 8 above)
Comment (dilipbiswal) [reply sql/core/src/test/scala/org/apache/spark/sql/test/DataFrameReaderWriterSuite.scala]: @cloud-fan I forgot :-) I will double check and get back.

File: docs/sql-programming-guide.md [hunk 10]
```diff
@@ -1930,6 +1930,8 @@ working with timestamps in `pandas_udf`s to get the best performance, see
     - Literal values used in SQL operations are converted to DECIMAL with the exact precision and scale needed by them.
     - The configuration `spark.sql.decimalOperations.allowPrecisionLoss` has been introduced. It defaults to `true`, which means the new behavior described here; if set to `false`, Spark uses previous rules, ie. it doesn't adjust the needed scale to represent the values and it returns NULL if an exact representation of the value is not possible.
 
+ - Since Spark 2.3, writing an empty dataframe to a directory launches at least one write task, even if physically the dataframe has no partition. This introduces a small behavior change that for self-describing file formats like Parquet and Orc, Spark creates a metadata-only file in the target directory when writing an empty dataframe, so that schema inference can still work if users read that directory later. The new behavior is more reasonable and more consistent regarding writing empty dataframe.
```
Comment (cloud-fan) [inline docs/sql-programming-guide.md]: `Spark creates a metadata-only file in the target directory when writing an empty dataframe` -> `0-partition dataframe`. We only get a behavior change if the dataframe has no partition.

Comment (dilipbiswal) [reply docs/sql-programming-guide.md]: ok.

File: sql/core/src/test/scala/org/apache/spark/sql/test/DataFrameReaderWriterSuite.scala (hunk 8 above)
Comment (dilipbiswal) [reply sql/core/src/test/scala/org/apache/spark/sql/test/DataFrameReaderWriterSuite.scala]: @cloud-fan It fails in the executor like this -
```SQL
org.apache.parquet.schema.InvalidSchemaException: Cannot write a schema with an empty group: message spark_schema {
}

	at org.apache.parquet.schema.TypeUtil$1.visit(TypeUtil.java:27)
	at org.apache.parquet.schema.TypeUtil$1.visit(TypeUtil.java:37)
	at org.apache.parquet.schema.MessageType.accept(MessageType.java:58)
	at org.apache.parquet.schema.TypeUtil.checkValidWriteSchema(TypeUtil.java:23)
	at org.apache.parquet.hadoop.ParquetFileWriter.<init>(ParquetFileWriter.java:225)
	at org.apache.parquet.hadoop.ParquetOutputFormat.getRecordWriter(ParquetOutputFormat.java:342)
	at org.apache.parquet.hadoop.ParquetOutputFormat.getRecordWriter(ParquetOutputFormat.java:302)
	at org.apache.spark.sql.execution.datasources.parquet.ParquetOutputWriter.<init>(ParquetOutputWriter.scala:37)
	at org.apache.spark.sql.execution.datasources.parquet.ParquetFileFormat$$anon$1.newInstance(ParquetFileFormat.scala:151)
	at org.apache.spark.sql.execution.datasources.FileFormatWriter$SingleDirectoryWriteTask.newOutputWriter(FileFormatWriter.scala:376)
	at org.apache.spark.sql.execution.datasources.FileFormatWriter$SingleDirectoryWriteTask.execute(FileFormatWriter.scala:387)
	at org.apache.spark.sql.execution.datasources.FileFormatWriter$$anonfun$org$apache$spark$sql$execution$datasources$FileFormatWriter$$executeTask$3.apply(FileFormatWriter.scala:278)
	at org.apache.spark.sql.execution.datasources.FileFormatWriter$$anonfun$org$apache$spark$sql$execution$datasources$FileFormatWriter$$executeTask$3.apply(FileFormatWriter.scala:276)
	at org.apache.spark.util.Utils$.tryWithSafeFinallyAndFailureCallbacks(Utils.scala:1411)
	at org.apache.spark.sql.execution.datasources.FileFormatWriter$.org$apache$spark$sql$execution$datasources$FileFormatWriter$$executeTask(FileFormatWriter.scala:281)
	at org.apache.spark.sql.execution.datasources.FileFormatWriter$$anonfun$write$1.apply(FileFormatWriter.scala:206)
	at org.apache.spark.sql.execution.datasources.FileFormatWriter$$anonfun$write$1.apply(FileFormatWriter.scala:205)
	at org.apache.spark.scheduler.ResultTask.runTask(ResultTask.scala:87)
	at org.apache.spark.scheduler.Task.run(Task.scala:109)
	at org.apache.spark.executor.Executor$TaskRunner.run(Executor.scala:345)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.
```

Comment (dilipbiswal) [reply sql/core/src/test/scala/org/apache/spark/sql/test/DataFrameReaderWriterSuite.scala]: Let me study the code to see how we can fail earlier.

Comment (cloud-fan) [reply sql/core/src/test/scala/org/apache/spark/sql/test/DataFrameReaderWriterSuite.scala]: Let's open a JIRA. We can fix it in another PR.

Comment (dilipbiswal) [reply sql/core/src/test/scala/org/apache/spark/sql/test/DataFrameReaderWriterSuite.scala]: @cloud-fan OK Wenchen.

**Comment by dilipbiswal** [comment]
@gatorsmile @cloud-fan Hello, is there anything pending for this ? The reason i ask is, for the other [PR] (https://github.com/apache/spark/pull/20579) , i just realised that the write code, does not produce any metadata only files for 0 partition case. Please let me know..

**Comment by cloud-fan** [comment]
Can we hold it for a while? If RC3 fails, let's merge this to 2.3 branch. If RC3 passes, we should only merge it to master.

**Comment by dilipbiswal** [comment]
@cloud-fan Got it Wenchen. Thanks for your reply. I will hold off on 20579 for a while till we get this in.

**Comment by cloud-fan** [comment]
ah sorry I missed this one. Let's retarget this to 2.4 and update the migration guide, thanks!

**Comment by cloud-fan** [comment]
retest this please

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/FileFormatWriter.scala [hunk 11]
```diff
@@ -202,7 +211,7 @@ object FileFormatWriter extends Logging {
             committer,
             iterator = iter)
         },
-        0 until rdd.partitions.length,
+        0 until rddWithNonEmptyPartitions.partitions.length,
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/FileFormatWriter.scala]: nit: `rddWithNonEmptyPartitions.partitions.indices`

**Comment by cloud-fan** [comment]
LGTM

**Comment by jiangxb1987** [comment]
LGTM

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/FileFormatWriter.scala (hunk 11 above)
Comment (dilipbiswal) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/FileFormatWriter.scala]: made the change.  Learnt a new thing today :-)

**Comment by cloud-fan** [comment]
thanks, merging to master!

**Comment by dilipbiswal** [comment]
@cloud-fan @jiangxb1987 Thank you very much !!

**Comment by HyukjinKwon** [comment]
late LGTM too.
